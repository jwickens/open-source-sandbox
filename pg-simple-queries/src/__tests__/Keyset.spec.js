/**
 * @flow
 */

const { Db } = require('../')
const path = require('path')

const OPTIONS = {
  schemaBase: 'keysettests',
  schemaVersion: 1.0,
  overrideEnv: true,
  sqlDirectory: path.join(__dirname, 'sql')
}

const db = new Db(OPTIONS)
let table
beforeAll(async () => {
  await db.sync()
  table = db.getTable('test')
})

afterAll(async () => {
  await db.close()
})

describe('Keyset', () => {
  test('can add a seek to keyset', () => {
    const keyset = table.getKeyset().seek({field: 'id', value: 1})
    expect(keyset.toSql()).toMatchSnapshot()
  })

  test('can seek before keyset', () => {
    const keyset = table.getKeyset().seek({field: 'id', value: 1}).before()
    expect(keyset.toSql()).toMatchSnapshot()
  })

  test('can add a seek with desc sort', () => {
    const keyset = table.getKeyset().seek({field: 'id', value: 1, desc: true}).after()
    expect(keyset.toSql()).toMatchSnapshot()
  })

  test('can add a filter in to keyset', () => {
    const keyset = table.getKeyset().in({field: 'id', values: [1, 2, 3]})
    expect(keyset.toSql()).toMatchSnapshot()
  })

  test('can add a filter intersect to keyset', () => {
    const keyset = table.getKeyset().intersect({field: 'tags', values: ['b']})
    expect(keyset.toSql()).toMatchSnapshot()
  })

  test('can add a filter where to keyset', () => {
    const keyset = table.getKeyset().where({field: 'id', value: 1})
    expect(keyset.toSql()).toMatchSnapshot()
  })

  test('can add a filter null to keyset', () => {
    const keyset = table.getKeyset().isNull({field: 'id'})
    expect(keyset.toSql()).toMatchSnapshot()
  })

  test('can add a filter not null to keyset', () => {
    const keyset = table.getKeyset().notNull({field: 'id'})
    expect(keyset.toSql()).toMatchSnapshot()
  })

  test('can add a raw where and order clauses to the keyset', () => {
    const keyset = table.getKeyset().rawWhere('id is null').rawOrder('id asc nulls last')
    expect(keyset.toSql()).toMatchSnapshot()
  })

  test('can convert a keyset into an opaque cursor and back (search)', () => {
    const keyset = table.getKeyset()
      .seek({field: 'id', value: 0, searchQuery: 'my search query'})
      .notNull({field: 'id'})
      .in({field: 'id', values: [1, 2]})
      .where({field: 'id', value: 1})
      .intersect({field: 'tags', values: ['a']})
    const cursor = keyset.toOpaqueCursor()
    expect(cursor).toMatchSnapshot()
    const newKeyset = table.getKeyset().fromOpaqueCursor(cursor)
    const firstSql = keyset.toSql()
    const secondSql = newKeyset.toSql()
    expect(firstSql).toEqual(secondSql)
    expect(firstSql).toMatchSnapshot()
  })

  test('can convert a keyset into an opaque cursor and back (basic)', () => {
    const keyset = table.getKeyset()
      .seek({field: 'id', value: 1, desc: true})
      .isNull({field: 'id'})
    const firstSql = keyset.toSql()
    const secondSql = keyset.clone().toSql()
    expect(firstSql).toEqual(secondSql)
    expect(firstSql).toMatchSnapshot()
  })

  test('can parse a graphql style cursor query', () => {
    const keyset = table.getKeyset()
      .seek({field: 'id', value: 1, desc: true})
      .isNull({field: 'id'})
    const from = keyset.toOpaqueCursor()
    const nextKeyset = table.getKeyset().continue({from, next: 1})
    const prevKeyset = table.getKeyset().continue({from, prev: 1})
    expect(nextKeyset.seekParams).toMatchObject(keyset.seekParams)
    expect(prevKeyset.seekParams).toMatchObject(keyset.seekParams)
  })

  test('updating from result does not affect the original keyset', () => {
    const keyset = table.getKeyset()
      .seek({field: 'id'})

    const newKeyset = keyset.newFromResult({row: {id: 1}})

    expect(keyset.seekParams[0].value).toBeUndefined()
    expect(newKeyset.seekParams[0].value).toBeDefined()
  })
})
