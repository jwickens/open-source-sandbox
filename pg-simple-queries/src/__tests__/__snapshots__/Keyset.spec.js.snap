// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyset can add a filter in to keyset 1`] = `
"
      SELECT null
      FROM test
      WHERE id IN ('1','2','3')
      
      
      LIMIT 20
    "
`;

exports[`Keyset can add a filter intersect to keyset 1`] = `
"
      SELECT null
      FROM test
      WHERE tags && ARRAY['b']
      
      
      LIMIT 20
    "
`;

exports[`Keyset can add a filter not null to keyset 1`] = `
"
      SELECT null
      FROM test
      WHERE id IS NOT NULL
      
      
      LIMIT 20
    "
`;

exports[`Keyset can add a filter null to keyset 1`] = `
"
      SELECT null
      FROM test
      WHERE id IS NULL
      
      
      LIMIT 20
    "
`;

exports[`Keyset can add a filter where to keyset 1`] = `
"
      SELECT null
      FROM test
      WHERE id = '1'
      
      
      LIMIT 20
    "
`;

exports[`Keyset can add a raw where and order clauses to the keyset 1`] = `
"
      SELECT null
      FROM test
      WHERE id is null
      
      ORDER BY id asc nulls last
      LIMIT 20
    "
`;

exports[`Keyset can add a seek to keyset 1`] = `
"
      SELECT null
      FROM test
      WHERE test.id is not null AND test.id > '1'
      
      ORDER BY test.id ASC
      LIMIT 20
    "
`;

exports[`Keyset can add a seek with desc sort 1`] = `
"
      SELECT null
      FROM test
      WHERE test.id is not null AND test.id < '1'
      
      ORDER BY test.id DESC
      LIMIT 20
    "
`;

exports[`Keyset can convert a keyset into an opaque cursor and back (basic) 1`] = `
"
      SELECT null
      FROM test
      WHERE test.id is not null AND test.id < '1' AND id IS NULL
      
      ORDER BY test.id DESC
      LIMIT 20
    "
`;

exports[`Keyset can convert a keyset into an opaque cursor and back (search) 1`] = `"eyJzIjpbeyJ2IjowLCJmIjoiaWQiLCJkIjpmYWxzZSwicSI6Im15IHNlYXJjaCBxdWVyeSJ9XSwiZiI6W3sidiI6WzEsMl0sImYiOiJpZCJ9XSwidyI6W3sidiI6MSwiZiI6ImlkIn1dLCJpIjpbeyJ2IjpbImEiXSwiZiI6InRhZ3MifV0sIm4iOlt7ImYiOiJpZCIsIm4iOmZhbHNlfV19"`;

exports[`Keyset can convert a keyset into an opaque cursor and back (search) 2`] = `
"
      SELECT similarity(test.id, 'my search query') as sml
      FROM test
      WHERE test.id is not null AND id like '%my search query%' AND similarity(test.id, 'my search query') > '0' AND id IN ('1','2') AND id = '1' AND tags && ARRAY['a'] AND id IS NOT NULL
      
      ORDER BY similarity(test.id, 'my search query') ASC
      LIMIT 20
    "
`;

exports[`Keyset can seek before keyset 1`] = `
"
      SELECT null
      FROM test
      WHERE test.id is not null AND test.id < '1'
      
      ORDER BY test.id ASC
      LIMIT 20
    "
`;
